docs/proposals/20251120-pub-support-podgroup.md

```diff
@@ -0,0 +1,200 @@
+---
+apiVersion: policy.kruise.io/v1alpha1
+kind: PodUnavailableBudget
+metadata:
+  name: llm-serving-budget
+  namespace: inference
+spec:
+  selector:
+    matchLabels:
+      app: llm-serving
+  maxUnavailable: 1         # interpreted as groups
+  podGroupPolicy:
+    groupLabelKey: leaderworkerset.sigs.k8s.io/group-index
+    minReadyReplicas: 3     # 1 leader + 2 workers Ready → group serves traffic
+```
+
+The PUB now ensures at most one shard goes offline at a time, aligning disruption budgets with the real service unit.
+
+#### API Definition
+
+```go
+type PodUnavailableBudgetSpec struct {
+    // existing fields: selector, targetReference, maxUnavailable, minAvailable, etc.
+
+    // PodGroupPolicy defines the policy for grouping pods and calculating
+    // availability at the group level. When nil, legacy per-pod semantics apply.
+    PodGroupPolicy *PodGroupPolicy `json:"podGroupPolicy,omitempty"`
+}
+
+type PodGroupPolicy struct {
+    // GroupLabelKey specifies the label key used to group Pods.
+    // Pods with the same value belong to the same group.
+    // Example: "leaderworkerset.sigs.k8s.io/group-index"
+    GroupLabelKey string `json:"groupLabelKey"`
+
+    // MinReadyReplicas is the minimum number of Ready pods required
+    // for a group to be considered Available.
+    // Default (when nil) is 1.
+    MinReadyReplicas *int32 `json:"minReadyReplicas,omitempty"`
+}
+```
+
+No status schema changes are required. When `podGroupPolicy` is set, the existing status fields (`currentAvailable`, `desiredAvailable`, `unavailableReplicas`) are interpreted as **group counts**.
+
+#### Semantics
+
+- **Grouping**: Pods are grouped by `GroupLabelKey`. Pods missing the label fall back to legacy per-Pod protection, and the PUB emits a warning condition.
+- **Group Availability**: A group is Available when `ReadyPodsInGroup >= MinReadyReplicas` (default 1). The Ready determination matches existing PUB logic (PodReady true and not already disrupted).
+- **Budget Interpretation**:
+  - `maxUnavailable=1` → “At most 1 group Unavailable.”
+  - `minAvailable=2` → “At least 2 groups Available.”
+  - Both fields can coexist, enforcing both constraints.
+- **Backward Compatibility**: With `podGroupPolicy` unset (default), PUB behaves exactly as today.
+
+#### Design Details
+
+- **Controller responsibilities**:
+  - List Pods selected by the PUB.
+  - Group them by `GroupLabelKey`.
+  - Compute Ready counts per group (respecting disruption annotations as today).
+  - Determine Available vs Unavailable groups using `MinReadyReplicas`.
+  - Update status counts with group numbers.
+  - Continue protecting Pods missing the label using legacy semantics and surface warnings.
+
+- **Status updates**:
+  - `currentAvailable = number of Available groups`.
+  - `desiredAvailable = groups required by minAvailable or maxUnavailable`.
+  - `unavailableReplicas = number of Unavailable groups`.
```

Comment:
currently status.UnavailablePods record the name of pod, if dealing with podgroup, what shall be recorded?

Reply:
A new field `stauts.UnavailablePodGroups` may be necessary.